src/Microsoft.AspNet.OData.Shared/TimeZoneInfoHelper.cs

```diff
@@ -9,18 +9,30 @@ internal class TimeZoneInfoHelper
     {
         private static TimeZoneInfo _defaultTimeZoneInfo;
 
+        private static object _timeZoneLock = new object();
+
         public static TimeZoneInfo TimeZone
         {
             get
             {
-                if (_defaultTimeZoneInfo == null)
+                var timeZone = TimeZoneInfo.Local;
```

Comment:
I am worry about the timezone lock.  We need to refactor this static.

I did the refactor at our 8.0. Maybe we can do it here.

Reply:
In 8.0 you refactored the code to move the `TimeZoneInfo` into an instance of `ODataOptions` so `SetTimeZoneInfo` becomes an instance method.

In WebApi, `SetTimeZoneInfo` is a public `static` method on `ODataRouteBuilderExtensions` (`IRouteBuilder`) and it is using `TimeZoneInfoHelper.TimeZone` to save the `TimeZoneInfo` in a static field. The difficulty of a **refactor** is in figuring out where to store a per-instance value through an extension method (extension properties don't exist). The `IRouteBuilder` has a size that is fixed at construction so space can't be added to it. The proper technique is through inheritance, but that will break the existing public API.

So I agree that it should be refactored as you have done in 8.0, but I don't see how to do that without going to WebAPI 8.0.

I agree that locks are _ugly_, but static fields in a static class are inherently not thread-safe (and strongly discouraged). So a lock is required.

However, the impact of the lock on real-world usage is trivial. Multi-threaded calls to Get/SetTimeZoneInfo is extremely rare, the common case is to set it only once during configuration.

Reply:
Is there any solution that we can resolve the problem without introducing lock?

Reply:
Maybe AsyncLocal<T>. It  essentially gives each thread its own copy and ensures that changes are propagated through the call chain even if thread context changes happen. 

```CSHARP
        private static AsyncLocal<TimeZoneInfo> _defaultTimeZoneInfo = ...;

        public static TimeZoneInfo TimeZone { get => _defaultTimeZoneInfo.Value; set => _defaultTimeZoneInfo.Value = value}
```

More research and more testing would be required to see if that achieves what is desired.

Reply:
@chrisspre `AsyncLocal` will not work, conceptually it is analogous to Thread Local Storage with the added property that the values "flow" across Task execution to another thread if during an `await` the thread is blocked and the flow has to continue on another thread. Internally it is using the `ExecutionContext` to store the "local" values. Once a thread's "flow" completes (Thread Abort or Join; ThreadPool completion) the AsyncLocal data is cleared.

So the semantics are not correct:

- Any `TimeZoneInfo` set and saved in an `AsyncLocal` cannot be read by any other thread unless it is a thread that "flowed" through Async from the original thread.
- Any `TimeZoneInfo` set and saved in an `AsyncLocal` will be cleared when the Thread exits or is returned to the ThreadPool. For a ASP.Net OData server that means the `TimeZoneInfo`  will only persist for the duration of the HttpRequest.
- The current semantics of `SetTimeZoneInfo` is that the value persists for the live of the service

The question @xuzhg  posed is can we avoid a lock? Underlying that question is the more fundamental question of **why a lock should be avoided**? 

Locks are _heavyweight_ (TLB shootdowns) but in this case it is written once and read many which is the optimal case.

In this case the `TimeZoneInfo` is rarely called (1-2 times per request), not in any critical path or loop.

In this case the lock's hold time is extremely short.

Reply:
@dxrdxr @xuzhg  @chrisspre  what if the tests that use the TimeZoneInfo were set to run sequentially?

In production, is the TimeZone changed per request, per route or once during the entire lifetime of the app? If it's meant to be set only once in production, then maybe it might be sufficient to let the tests run sequentially and avoid the locks in production. What do you think?

Reply:
@habbes I have put all the `TimeZoneInfo` tests in the same "Collection" so that the do now run sequentially. But that is necessary even with the locked access. The problem is that 2 tests running in parallel one sets it to UTC-8 and another one sets it to UTC, then the first accesses it getting UTC and fails the test.

I have done some deeper C#/CLR research. Based on [Atomicity of variable references](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/variables#atomicity-of-variable-references) "Reads and writes of the following data types are atomic: [...] reference types." I mistakenly thought that a `TimeZoneInfo` was a structure type (like `DateTime`) which is a Value type and not atomic.

So I agree and withdraw the lock changes.